README.md
<h1 align="center">Flask-Gopher</h1>

<p align="center">A <a href="http://flask.pocoo.org/">Flask</a> extension to support the <a href="https://en.wikipedia.org/wiki/Gopher_(protocol)">Gopher</a> protocol.</p>

<p align="center">
  <img alt="gopher" src="resources/gopher_alt.jpg"/>
</p>

<p align="center">
  <a href="https://pypi.python.org/pypi/flask-gopher/">
    <img alt="pypi" src="https://img.shields.io/pypi/v/flask-gopher.svg?label=version"/>
  </a>
  <a href="https://pypi.python.org/pypi/flask-gopher/">
    <img alt="python" src="https://img.shields.io/badge/python-3.7+-blue.svg"/>
  </a>
  <a href="https://github.com/michael-lazar/flask-gopher/actions">
    <img alt="tests" src="https://github.com/michael-lazar/flask-gopher/workflows/Test/badge.svg"/>
  </a>
</p>

## Contents

* [Demo](#demo)
* [About](#about)
* [Quickstart](#quickstart)
* [Installation](#installation)
* [Building Gopher Menus](#building-gopher-menus)
* [Using Templates](#using-templates)
* [Gopher and WSGI](#gopher-and-wsgi)
* [Gopher Protocol References](#gopher-protocol-references)

## Demo

A live demonstration of the Flask-Gopher server is available in
gopherspace at the following URL:

---

<p align="center">
<b><a href="gopher://mozz.us:7005">gopher://mozz.us:7005</a></b><br>
</p>

---

## About

*What is gopher?*

Gopher is an alternative to the World Wide Web that peaked in
popularity in the early 90's. There are still a handful of gopher
sites maintained by enthusiasts; you can learn more about its history
at [floodgap](http://gopher.floodgap.com/gopher/).

*What is flask-gopher?*

Flask-Gopher is a Flask extension that adds a thin *Gopher -> HTTP*
compatability layer around the built-in webserver. It allows you to
build fully [RFC 1466](https://tools.ietf.org/html/rfc1466) compliant
gopher servers, with complete access to Flask's routing, templating
engine, debugger, and more!

*Who is this for?*

I created this extension because I wanted to experiment with building
dynamic gopher applications, and I felt limited by the lack of
flexibility in other gopher servers. The target audience is web
developers with experience using a high level web framework like
Django or Ruby on Rails. You should feel comfortable writing python
code and cross-referencing the official Flask documentation.

## Quickstart

```python
from flask import Flask, url_for
from flask_gopher import GopherExtension, GopherRequestHandler

app = Flask(__name__)
gopher = GopherExtension(app)

@app.route('/')
def index():
    return gopher.render_menu(
        gopher.menu.title('My GopherHole'),
        gopher.menu.dir('Home', url_for('index')),
        gopher.menu.info("Look Ma, it's a gopher server!"))

if __name__ == '__main__':
   app.run('127.0.0.1', 70, request_handler=GopherRequestHandler)
```

## Installation

This package requires **Python v3.7 or higher**

```
pip install flask_gopher
```

## Building Gopher Menus

Gopher menus are structured text files that display information
about the current page and contain links to other gopher resources.
A gopher menu is loosely equivalent to an HTML document with only
``<a>`` and ``<span>`` tags. Each line in the menu has a *type*
that describes what kind of resource it links to (text, binary, html,
telnet, etc.).

Flask-Gopher provides several helper methods for constructing gopher
menu lines:

| Method         | Link Descriptor | Meaning                                 |
|----------------|-----------------|-----------------------------------------|
| menu.text      | 0               | Plain text file                         |
| menu.dir       | 1               | Gopher menu                             |
| menu.ccso      | 2               | CCSO database; other databases          |
| menu.error     | 3               | Error message                           |
| menu.binhex    | 4               | Macintosh BinHex file                   |
| menu.archive   | 5               | Archive file (zip, tar, gzip)           |
| menu.uuencoded | 6               | UUEncoded file                          |
| menu.query     | 7               | Search query                            |
| menu.telnet    | 8               | Telnet session                          |
| menu.bin       | 9               | Binary file                             |
| menu.gif       | g               | GIF format graphics file                |
| menu.image     | I               | Other Image file                        |
| menu.doc       | d               | Word processing document (ps, pdf, doc) |
| menu.sound     | s               | Sound file                              |
| menu.video     | ;               | Video file                              |
| menu.info      | i               | Information line                        |
| menu.title     | i               | Title line                              |
| menu.html      | h               | HTML document                           |

Most of these methods require a text description for the link, and
will accept a path selector and a host/port. They return a line of
text that has been pre-formatted for a gopher menu. You can then pass
all of the lines along into ``gopher.render_menu()`` to build the
response body.

```python
@app.route('/')
def index():
    return gopher.render_menu(
        # Link to an internal gopher menu
        gopher.menu.dir('Home', '/'),

        # Link to an external gopher menu
        gopher.menu.dir('XKCD comics', '/fun/xkcd', host='gopher.floodgap.com', port=70),

        # Link to a static file, using flask.url_for() to build a relative path
        gopher.menu.image('Picture of a cat', url_for('static', filename='cat.png')),

        # Link to an external web page
        gopher.menu.html('Project source', 'https://github.com/michael-lazar/flask-gopher'),

        # A text info line
        gopher.menu.info('This is informational text'),

        # Plain text will be converted into info lines
        "\n    There's no place\n    like ::1\n",

        # You can also format your links manually
        "0About this page\t/about.txt\t127.0.0.1\t8007")
```

Here's what the rendered menu looks like as plain text:

```
$ curl gopher://localhost:8007
1Home	/	127.0.0.1	8007
1XKCD comics	/fun/xkcd	gopher.floodgap.com	70
IPicture of a cat	/static/cat.png	127.0.0.1	8007
hProject source	URL:https://github.com/michael-lazar/flask-gopher	127.0.0.1	8007
iThis is informational text	fake	example.com	0
i 	fake	example.com	0
i    There's no place	fake	example.com	0
i    like ::1	fake	example.com	0
i 	fake	example.com	0
0About this page	/about.txt	127.0.0.1	8007
```

And here's what it looks like inside of a gopher client:

<p align="center">
  <img alt="gopher" src="resources/example_menu.png"/>
</p>

## Using Templates

You can use Flask's Jinja2 templating engine to layout gopher menus.
Flask-Gopher attaches ``gopher`` to the template namespace so you can
access the menu helper functions. The recommended naming convention
for gopher template files is to add a *.gopher* suffix. An example
template file is shown below:

**templates/example_menu.gopher**

```
{{ 'Centered Title' | underline('-') | center }}

{{ gopher.menu.dir('Home', url_for('index')) }}

Hello from my gopher template!
Your IP address is {{ request.remote_addr }}

{{ '_' * gopher.width }}
{{ ('Served by ' + request.environ['SERVER_SOFTWARE']) | rjust }}
```

Call ``gopher.render_menu_template()`` from inside of your route to
compile the template into a gopher menu.

```python
@app.route('/')
def index():
    return gopher.render_menu_template('example_menu.gopher')
```

<p align="center">
  <img alt="gopher" src="resources/example_template.png"/>
</p>

## Gopher and WSGI

Python's WSGI (Web Server Gateway Interface) is an established API
that defines how python web servers (gunicorn, mod_wsgi, etc)
communicate with application frameworks (Flask, Django, etc). It
defines a clean boundary between low-level socket and request
handling, and high-level application logic.

WSGI was designed to be a very simple and flexible API, but at its
heart it's built around HTTP requests. As such, it incorperates some
HTTP specific components like request/response headers and status
codes. Gopher is more simplistic and doesn't use these components.
Here's an example of the difference in fetching a document with the
two protocols:

<table>
<tr><th colspan=2>HTTP</th><th colspan=2>Gopher</th></tr>
<tr><th>request</th><th>response</th><th>request</th><th>response</th></tr>
<tr>
<td width="20%"><pre>
GET /path HTTP/1.1
Accept: text/plain
Accept-Charset: utf-8
...more headers
</pre></td>
<td width="20%"><pre>
HTTP/1.1 200 OK
Server: Apache
Content-Type: text/html
...more headers<br>
(body)
</pre></td>
<td width="20%"><pre>/path\r\n</pre></td>
<td width="20%"><pre>(body)</pre></td>
</tr></table>

In order to resolve the differences between gopher and HTTP, *
*Flask-Gopher** implements a custom ``GopherRequestHandler``. The
handler hooks into the WSGI server (``werkzeug.BaseWSGIServer``). It
reads the first line of every TCP connection and determines which
protocol the client is attempting to use. If the client is using
gopher, the following assumptions are made:

- Set the request's *REQUEST_METHOD* to ``GET``
- Set the request's *SERVER_PROTOCOL* (e.g. *HTTP/1.1*) to ``gopher``
- Set the request's *wsgi.url_scheme* (e.g. *https*)  to ``gopher``
- Discard the response status line
- Discard all response headers

Doing this makes a gopher connection *appear* like a normal HTTP
request from the perspective of the WSGI application. It also provides
metadata hooks that can be accessed from the Flask request. For
example, you can respond the the request differently depending on
which protocol is being used:

```python
@app.route('/')
def index():
    if flask.request.scheme == 'gopher':
        return "iThis was a gopher request\tfake\texample.com\t0\r\n"
    else:
        return "<html><body>This was an HTTP request</body></html>"
```

## Gopher Protocol References

- https://tools.ietf.org/html/rfc1436 (1993)
- https://tools.ietf.org/html/rfc4266 (2005)
- https://tools.ietf.org/html/draft-matavka-gopher-ii-03 (2015)
- https://www.w3.org/Addressing/URL/4_1_Gopher+.html

An interesting side note, the python standard library used to contain
its own gopher module. It was deprecated in 2.5, and removed in
2.6. (<em>https://www.python.org/dev/peps/pep-0004/</em>)


>     Module name:   gopherlib
>     Rationale:     The gopher protocol is not in active use anymore.
>     Date:          1-Oct-2000.
>     Documentation: Documented as deprecated since Python 2.5.  Removed
>                    in Python 2.6.

A reference gopher client still exists in the old python SVN
trunk: https://svn.python.org/projects/python/trunk/Demo/sockets/gopher.py


CHANGELOG.md
## v3.0.0 (2023-02-15)

This version pulls in some long overdue dependency updates and adds
support for the latest versions of Flask and Python.

- Supported Python versions: `3.7`, `3.8`, `3.9`, `3.10`, `3.11`.
- Supported Flask versions: `2.1`, `2.2`

This version also removes the capability to negotiate TLS over
gopher. The code for this was particularly annoying to monkey-patch,
and the feature never gained traction to make it worth maintaining.

- Removed the `make_gopher_ssl_server` function.
- Removed the following WSGI server classes:
  - ``GopherBaseWSGIServer``
  - ``GopherSimpleWSGIServer``
  - ``GopherThreadedWSGIServer``
  - ``GopherForkingWSGIServer``

### v2.2.1 (2020-04-11)

- Pin the werkzeug dependency to avoid breaking dependency changes.

### v2.2.0 (2020-01-11)

- Added support for python 3.8
- Dropped support for python 3.4
- Added support for flask 1.1
- Renamed the following methods:

  - ``menu.submenu()`` -> ``menu.dir()``
  - ``menu.file()`` -> ``menu.text()``
  - ``menu.binary()`` -> ``menu.bin()``

This was done as a personal preference because I found the original method
names to be overly verbose and hard to remember. The original methods will
remain for backwards compatibility, but they will no longer be listed in
the documentation.

### v2.1.1 (2019-04-10)

- Add a helper method to generate a gopher menu from a directory.
- Fixed bug when redirecting to HTTP pages that was causing query parameters
  to get lost.

### v2.1.0 (2019-01-08)

- Added support for establishing TLS connections with gopher clients. See the
  documentation for more details.

### v2.0.0 (2018-01-17)

- Major restructuring of the codebase and the public API.
- Added a ``TextFormatter`` class with several helper methods for formatting
  ASCII.
- Added a ``GopherSessionInterface`` class in order to support flask sessions.
- Added the ability to set the ``SERVER_NAME`` to an external URL.
- Included a ``demo/`` directory with a complete example gopher server.
- Several minor bug fixes and improvements.

### v1.0.0 (2018-01-07)

- First official release


tests/test_package.py
import os
import socket
import unittest
from threading import Thread
from urllib.request import Request, urlopen

from werkzeug.serving import make_server
from flask import Flask, request, url_for
from flask_gopher import GopherExtension, GopherRequestHandler, GopherMenu
from flask_gopher import render_menu, render_menu_template


TEST_DIR = os.path.dirname(os.path.realpath(__file__))


class TestFunctional(unittest.TestCase):
    """
    Because the flask_gopher extension is built on hacking the WSGI protocol
    and the werkzeug HTTP server, I feel that the only way to sincerely
    know that it's working properly is through functional testing.

    So this class will spin up a complete test flask application and serve
    it on a local TCP port in a new thread. The tests will send real gopher
    connection strings to the server and check the validity of the response
    body from end-to-end.
    """
    app = gopher = server = thread = None

    @classmethod
    def setUpClass(cls):
        """
        Spin up a fully-functional test gopher application in a new thread.
        """

        cls.app = app = Flask(__name__, template_folder=TEST_DIR)
        cls.gopher = gopher = GopherExtension(app)

        @app.route('/')
        def home():
            return 'Hello World!'

        @app.route('/menu')
        def menu():
            return render_menu(
                gopher.submenu('Submenu', url_for('menu')),
                gopher.file('Text', url_for('static', filename='file.txt')),
                gopher.info('x' * 200),  # Long lines should be truncated
                'x' * 200,  # Should looks the same as using menu.info()
                gopher.info('foo\r\nbar\t'))  # Tabs and newlines should be stripped

        @app.route('/template')
        def template():
            return render_menu_template('example_template', items=range(3))

        @app.route('/page/<int:page>')
        def page(page):
            return 'page: ' + str(page)

        @app.route('/internal_url')
        def internal_url():
            return url_for('page', page=2)

        @app.route('/external_menu_url')
        def external_menu_url():
            return gopher.url_for('page', page=2, _external=True)

        @app.route('/search')
        def search():
            return 'query: ' + request.environ['SEARCH_TEXT']

        @app.route('/internal_error')
        def internal_error():
            return 1 // 0

        # This is the same thing as calling app.run(), but it returns a handle
        # to the server so we can call server.shutdown() later.
        cls.server = make_server('127.0.0.1', 0, app, request_handler=GopherRequestHandler)
        cls.thread = Thread(target=cls.server.serve_forever)
        cls.thread.start()

    @classmethod
    def tearDownClass(cls):
        """
        Try to kill the server.
        """
        cls.server.shutdown()
        cls.thread.join(timeout=5)

    @classmethod
    def send_data(cls, data):
        """
        Send byte data to the server using a TCP/IP socket.
        """
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.connect((cls.server.host, cls.server.port))
            s.sendall(data)

            # Receive data until the server disconnects
            chunks = []
            while True:
                data = s.recv(2048)
                if not data:
                    break
                chunks.append(data)
            return b''.join(chunks)

    def test_empty_request(self):
        """
        Empty request strings should return the root url.
        """
        resp = self.send_data(b'\n')
        self.assertEqual(resp, b'Hello World!')
        resp = self.send_data(b'\r\n')
        self.assertEqual(resp, b'Hello World!')

    def test_root_url(self):
        resp = self.send_data(b'/\r\n')
        self.assertEqual(resp, b'Hello World!')

    def test_multiline_request(self):
        """
        Anything past the first line of the request should be ignored.
        """
        resp = self.send_data(b'/\r\ntwo line request\r\nthird line')
        self.assertEqual(resp, b'Hello World!')

    def test_gopher_plus(self):
        """
        Gopher+ connections should ignore anything after the second tab.
        """
        resp = self.send_data(b'/\t\tgopher plus data\r\n')
        self.assertEqual(resp, b'Hello World!')

    def test_search(self):
        """
        A tab in the request line should indicate search query text.
        """
        resp = self.send_data(b'/search\r\n')
        self.assertEqual(resp, b'query: ')
        resp = self.send_data(b'/search\tMy Search String\r\n')
        self.assertEqual(resp, b'query: My Search String')

    def test_not_found(self):
        """
        Check that the default 404 routing returns a gopher menu response.
        """
        resp = self.send_data(b'/invalid_url\r\n')
        self.assertTrue(resp.startswith(b'iError: 404 Not Found\tfake\texample.com\t0\r\n'))
        self.assertIn(b'The requested URL was not found', resp)
        self.assertTrue(resp.endswith(b'\r\n.\r\n'))

    def test_internal_error(self):
        """
        Check that internal errors return a gopher menu response.
        """
        resp = self.send_data(b'/internal_error\r\n')
        self.assertTrue(resp.startswith(b'iError: 500 Internal Error\tfake\texample.com\t0\r\n'))
        self.assertTrue(resp.endswith(b'\r\n.\r\n'))

    def test_internal_url(self):
        """
        Internal links should work with the default flask url_for method.
        """
        resp = self.send_data(b'/internal_url\r\n')
        self.assertEqual(resp, b'/page/2')

    def test_external_menu_url(self):
        """
        External Gopher menu links should have a "/1/" at the root.
        """
        resp = self.send_data(b'/external_menu_url\r\n')
        url = 'gopher://%s:%s/1/page/2' % (self.server.host, self.server.port)
        self.assertEqual(resp, url.encode())

    def test_url_redirect(self):
        """
        Selectors starting with URL:<path> should be sent an HTML document.
        """
        resp = self.send_data(b'/URL:https://gopher.floodgap.com\r\n')
        self.assertTrue(resp.startswith(b'<HTML>'))
        self.assertTrue(resp.endswith(b'</HTML>'))

        href = b'<A HREF="https://gopher.floodgap.com">https://gopher.floodgap.com</A>'
        self.assertIn(href, resp)

    def test_http_get(self):
        """
        Regular HTTP requests should still work and headers should be passed
        """
        url = 'http://%s:%s/' % (self.server.host, self.server.port)
        request = Request(url)
        resp = urlopen(request)
        self.assertEqual(resp.status, 200)
        self.assertIn('Content-Type', resp.headers)
        self.assertIn('Content-Length', resp.headers)
        self.assertTrue(resp.headers['Server'].startswith('Flask-Gopher'))
        self.assertEqual(resp.read(), b'Hello World!')

    def test_render_menu(self):
        """
        Check that gopher menu pages are properly formatted.
        """
        resp = self.send_data(b'/menu\r\n')
        text = '\r\n'.join([
            '1Submenu\t/menu\t{host}\t{port}',
            '0Text\t/static/file.txt\t{host}\t{port}',
            'i{line}\tfake\texample.com\t0',
            'i{line}\tfake\texample.com\t0',
            'ifoobar\tfake\texample.com\t0',
            '.', ''])
        text = text.format(
            host=self.server.host,
            port=self.server.port,
            line='x' * self.gopher.width)
        self.assertEqual(resp, text.encode())

    def test_render_menu_template(self):
        """
        Check that jinja templates generate correctly for gopher menus.
        """
        resp = self.send_data(b'/template\r\n')
        text = '\r\n'.join([
            'i{space}Title\tfake\texample.com\t0',
            'igopher://{host}:{port}/1/menu\tfake\texample.com\t0',
            '1Home\t/menu\t{host}\t{port}',
            'i0\tfake\texample.com\t0',
            'i1\tfake\texample.com\t0',
            'i2\tfake\texample.com\t0',
            '.', ''])
        text = text.format(
            host=self.server.host,
            port=self.server.port,
            space=' ' * (self.gopher.width - len('Title')))
        self.assertEqual(resp, text.encode())


class TestGopherMenu(unittest.TestCase):
    def setUp(self):
        self.menu = GopherMenu('10.10.10.10', 7007)

    def test_default_entry(self):
        line = self.menu.file('Hello World')
        self.assertEqual(line, '0Hello World\t/\t10.10.10.10\t7007')

    def test_default_no_port_external(self):
        line = self.menu.submenu('Hello World', host='hngopher.com')
        self.assertEqual(line, '1Hello World\t/\thngopher.com\t70')

    def test_strip_invalid_characters(self):
        line = self.menu.query('Hello \r\nWorld\t', selector='/\tfoo\t')
        self.assertEqual(line, '7Hello World\t/foo\t10.10.10.10\t7007')

    def test_html(self):
        line = self.menu.html('Firefox', 'http://www.firefox.com')
        self.assertEqual(line, 'hFirefox\tURL:http://www.firefox.com\t10.10.10.10\t7007')

    def test_title(self):
        line = self.menu.title('Hello World')
        self.assertEqual(line, 'iHello World\tTITLE\texample.com\t0')


if __name__ == '__main__':
    unittest.main()


demo/run_server.py
#!/usr/bin/env python3
"""
This is a demo application built using Flask-Gopher.

Perhaps the easiest way to get up and running is:

    $ git clone https://github.com/michael-lazar/flask-gopher.git
    $ cd flask-gopher
    $ python3 -m virtualenv venv
    $ source venv/bin/activate
    $ pip install .
    $ ./demo/run_server.py

Then, connect to the server with:

    $ lynx gopher://localhost:7005
"""
import json
import os
from collections import OrderedDict

from flask import Flask, abort, request, session, url_for
from pyfiglet import FigletFont
from tabulate import tabulate_formats

from flask_gopher import GopherExtension, GopherRequestHandler

ROOT_DIR = os.path.dirname(__file__)

app = Flask(__name__, static_url_path="")
app.config.from_pyfile("demo.cfg")

gopher = GopherExtension(app)


@app.route("/")
def index():
    return gopher.render_menu_template("index.gopher")


@app.route("/render-figlet")
def figlet():
    """
    Render the search text using all possible figlet fonts.

    This loops through hundreds of fonts and may take a few seconds to return.
    """
    text = request.environ["SEARCH_TEXT"]
    lines = []
    for font in sorted(FigletFont.getFonts()):
        lines.extend(gopher.formatter.figlet(text, font=font).splitlines())
        lines.append(font)
        lines.append("")
    return gopher.render_menu(*lines)


@app.route("/render-tables")
def tables():
    """
    Render an ascii table using all possible styles provided by tabulate.
    """
    data = [
        ("text", "int", "float"),
        ("Cell 1", 100, 1 / 3),
        ("Cell 2", -25, 0.001),
        ("Cell 3", 0, 0),
    ]
    lines = []
    for table_fmt in tabulate_formats:
        lines.append(table_fmt)
        lines.append("")
        lines.append(gopher.formatter.tabulate(data, "firstrow", table_fmt))
        lines.append("\n")
    return gopher.render_menu(*lines)


@app.route("/error/<int:code>/<path:path>")
def error(code, path):
    """
    Endpoint that simulates error behavior for different types of files.
    """
    if code == 500:
        # Will raise a divide by zero error with a stack trace
        return 1 // 0
    else:
        # Return the specified error code through flask
        abort(code)


@app.route("/demo-links")
def demo_links():
    return gopher.render_menu_template("demo_links.gopher")


# This example uses the app's static folder, but you can also point to any
# absolute filepath that you want to serve files from.
static_directory = gopher.serve_directory(app.static_folder, "demo_directory", show_timestamp=True)


@app.route("/demo-directory")
@app.route("/demo-directory/<path:filename>")
def demo_directory(filename=""):
    is_directory, file = static_directory.load_file(filename)
    if is_directory:
        return gopher.render_menu_template("demo_directory.gopher", filename=filename, listing=file)
    else:
        return file


@app.route("/demo-errors")
def demo_errors():
    return gopher.render_menu_template("demo_errors.gopher")


@app.route("/demo-session", defaults={"action": None})
@app.route("/demo-session/<action>")
def demo_session(action):
    if action == "create":
        session["id"] = request.environ["SEARCH_TEXT"]
    elif action == "delete":
        session.clear()
    return gopher.render_menu_template("demo_session.gopher", action=action)


@app.route("/demo-formatting")
def demo_formatting():
    table_data = [
        ("text", "int", "float"),
        ("Cell 1", 100, 1 / 3),
        ("Cell 2", -25, 0.001),
        ("Cell 3", 0, 0),
    ]
    return gopher.render_menu_template("demo_formatting.gopher", table_data=table_data)


@app.route("/demo-environ")
def demo_environ():
    environ_table = [("Field", "Value")]
    for key, val in sorted(request.environ.items()):
        if not key.startswith(("werkzeug", "wsgi")):
            environ_table.append((key, val))

    return gopher.render_menu_template("demo_environ.gopher", environ_table=environ_table)


@app.route("/demo-form", defaults={"field": None})
@app.route("/demo-form/<field>")
def demo_form(field):
    form_fields = OrderedDict(
        [
            ("first_name", "First name"),
            ("last_name", "Last name"),
            ("email", "Email"),
            ("phone", "Phone number"),
            ("address", "Address"),
        ]
    )

    # Check if there was a new field added to the request
    request_query = request.args.to_dict()

    if field in form_fields:
        request_query[field] = request.environ["SEARCH_TEXT"]

    # Build the form using the currently populated fields
    lines = []
    for name, description in form_fields.items():
        if name in request_query:
            lines.append(f"{description:<13}: {request_query[name]}")
        else:
            url = url_for("demo_form", _external=False, field=name, **request_query)
            lines.append(gopher.menu.query(f"{description:<13}:", url))

    # Add the buttons at the bottom of the form
    lines.append("")
    if request_query:
        lines.append(gopher.menu.dir("clear", url_for("demo_form")))
    else:
        lines.append("clear")
    if request_query.keys() == form_fields.keys():
        url = url_for("demo_form", _external=False, field="submit", **request_query)
        lines.append(gopher.menu.dir("submit", url))
    else:
        lines.append("submit")

    # Show the query data if the "submit" button was pressed
    if field == "submit":
        lines.append("")
        lines.append("Submitted form data:")
        lines.append(json.dumps(request_query, indent=4, sort_keys=True))

    form = "\r\n".join(lines)
    return gopher.render_menu_template("demo_form.gopher", form=form)


if __name__ == "__main__":
    app.run(
        host=app.config["HOST"],
        port=app.config["PORT"],
        threaded=app.config["THREADED"],
        processes=app.config["PROCESSES"],
        request_handler=GopherRequestHandler,
    )
